### ISSUE TITLE
Add events to models

### ISSUE THREAD

**Issue by filip-jezek** [creative-connections/Bodylight.js-Composer#22]
A JS code, run on model event (initial, start, stop, step, crash .. ?), would enable to hand-code a number of features, which are not yet introduced into the composer GUI.

**Comment by david-polak** [comment]
events: [
    'play',
    'pause',
    'reset',
    'tick',
    'beforeEnterInitializationMode',
    'enterInitializationMode',
    'termination',
  ],

**Comment by david-polak** [comment]
'play',
    'pause',
    'reset',
    'tick',
    'terminate',
    'beforeEnterInitializationMode',
    'enterInitializationMode',

**Comment by filip-jezek** [comment]
+ 'warning' (from assert)?

**Comment by filip-jezek** [comment]
"Regarding the Tank overflow; The real behavior of the Test bench is to stop the pump before it overflows. Would it be possible to set the voltage to 0V, if the water level reaches 0.6 meters?"

Model event OnTick would enable to check for the level and adjust the parameter (which is otherwise input from the slider) accordingly.

**Comment by filip-jezek** [comment]
And what about including pre-set (empty) event, which is called on each:

- page load
- model init
- model tick

This would enable initialization scripts (e.g. setting intiial dataset visibility false) and scripts observing some model value (e.g. value is over some threshold -> set input range to zero).

**Comment by david-polak** [comment]
This can be implemented in several ways. But backwards compatibility is gonna suffer in the future, once the model gets run in a webworker, any previous .bjp files containing JavaScript crossing the model/component boundary would break hard. 

Any code attached to a model method which would need to communicate with other components would have to use the web worker messaging system and message catching code would need to be written for the target component.

**Comment by filip-jezek** [comment]
Hm.. good point.. So any workaround on how to set the toggles true / false and hide datasets from graph on page load? This does not communicate with the model, at least.

**Comment by david-polak** [comment]
I don't believe it's going to have a large impact, since the plan for the model thread separation was to support both encapsulated and traditional model. The message layer between widgets and models should support both types of model endpoints.

This decision should be documented somewhere, I will go on with the implementation of this issue.
